packager/certcache/certcache.go

```diff
@@ -233,20 +319,63 @@ func (this *CertCache) isHealthy(ocspResp []byte) error {
 }
 
 // Returns the OCSP response and expiry, refreshing if necessary.
-func (this *CertCache) readOCSP() ([]byte, time.Time, error) {
+// TODO(banaag): Per twifkak's suggestion, consider:
+// It may also be interesting to try to separate the retry logic from the fetch logic. One approach comes to mind:
+//
+// retryWithBackoff(func() {
+//   ocsp, err := ...
+//   return true if successful
+// }, initialWaitTime, maxTries)
+func (this *CertCache) readOCSP(allowRetries bool) ([]byte, time.Time, error) {
 	var ocspUpdateAfter time.Time
-	ocsp, err := this.ocspFile.Read(context.Background(), this.shouldUpdateOCSP, func(orig []byte) []byte {
-		return this.fetchOCSP(orig, &ocspUpdateAfter)
-	})
-	if err != nil {
-		return nil, time.Time{}, errors.Wrap(err, "Updating OCSP cache")
-	}
-	if len(ocsp) == 0 {
-		return nil, time.Time{}, errors.New("Missing OCSP response.")
+	var err error
+	var maxTries int
+
+	ocsp := []byte(nil)
+	waitTimeInMinutes := 1
+	if !allowRetries || this.certFetcher == nil {
+		// If certFetcher is nil, that means we are not auto-renewing so don't retry OCSP.
+		maxTries = 1
+	} else {
+		maxTries = maxOCSPTries
 	}
-	err = this.isHealthy(ocsp)
-	if err != nil {
-		return nil, time.Time{}, errors.Wrap(err, "OCSP failed health check.")
+
+	for numTries := 0; numTries < maxTries; {
+		this.certsMu.RLock()
+		defer this.certsMu.RUnlock()
+		ocsp, err = this.ocspFile.Read(context.Background(), this.shouldUpdateOCSP, func(orig []byte) []byte {
+			return this.fetchOCSP(orig, this.certs, &ocspUpdateAfter, numTries > 0)
+		})
+		if err != nil {
+			if numTries >= maxTries-1 {
+				return nil, time.Time{}, errors.Wrap(err, "Updating OCSP cache")
+			} else {
+				numTries++
+				waitTimeInMinutes = waitForSpecifiedTime(waitTimeInMinutes, numTries)
+				continue
+			}
+		}
+		if len(ocsp) == 0 {
+			if numTries >= maxTries-1 {
+				return nil, time.Time{}, errors.New("Missing OCSP response.")
+			} else {
+				numTries++
+				waitTimeInMinutes = waitForSpecifiedTime(waitTimeInMinutes, numTries)
+				continue
+			}
+		}
+		if err := this.isHealthy(ocsp); err != nil {
+			if numTries >= maxTries-1 {
+				return nil, time.Time{}, errors.Wrap(err, "OCSP failed health check.")
+			} else {
+				numTries++
+				waitTimeInMinutes = waitForSpecifiedTime(waitTimeInMinutes, numTries)
+				continue
+			}
+		} else {
+			break
+		}
+		numTries++
```

Comment:
Is this reachable?

The previous condition block starting l.367 always exists I think.

Reply:
Fixed with refactor from comment above.